File: java-fxml/src/sample/Controller.java
package sample;

import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.scene.control.Label;

public class Controller {

    @FXML
    Label lbl_count;

    int _count = 0;

    @FXML
    public void initialize() {
        _count = 0;
        lbl_count.setText(Integer.toString(this._count));
    }

    @FXML
    public void onBtnUp(ActionEvent evt) {
        this._count += 1;
        lbl_count.setText(Integer.toString(this._count));
    }

    @FXML
    public void onBtnDown(ActionEvent evt) {
        this._count -= 1;
        lbl_count.setText(Integer.toString(this._count));
    }
}

File: deftype-fxml/src/deftype_fxml/controller.clj
(ns deftype-fxml.controller
  (:import
   [java.net URL]
   [java.util ResourceBundle]
   [javafx.fxml FXMLLoader FXML]
   [javafx.event ActionEvent EventHandler]))


(defn helloworld []
  (println "hihihi"))


;; works.. but Are we need this interface for just for `one` controller?
(definterface IController
  (^{FXML [] :tag void} initialize [])
  (^{FXML [] :tag void} onBtnUp [^javafx.event.ActionEvent event])
  (^{FXML [] :tag void} onBtnDown [^javafx.event.ActionEvent event]))


(deftype Controller
    [state
     ^{FXML [] :unsynchronized-mutable true} lbl_count]

  IController
  (^{FXML [] :tag void} initialize [this]
   (.setText lbl_count (str (:count @state))))

  (^{FXML [] :tag void} onBtnUp [this ^javafx.event.ActionEvent event]
   (swap! state update-in [:count] inc)
   (.setText lbl_count (str (:count @state))))

  (^{FXML [] :tag void} onBtnDown [this ^javafx.event.ActionEvent event]
   (swap! state update-in [:count] dec)
   (helloworld)
   (.setText lbl_count (str (:count @state)))))


(defn new-controller []
  (->Controller
   (atom {:count 0})
   nil ;; lbl_count
   ))
